File: PaperConnect/src/com/paperconnect/client/PaperConnect.java
package com.paperconnect.client;

import java.util.ArrayList;

import com.google.gwt.core.client.EntryPoint;
import com.google.gwt.core.client.GWT;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONParser;
import com.google.gwt.json.client.JSONString;
import com.google.gwt.json.client.JSONValue;
import com.google.gwt.user.client.rpc.AsyncCallback;
import com.google.gwt.user.client.ui.Label;
import com.google.gwt.user.client.ui.RootPanel;
import com.paperconnect.exception.InvalidIdException;
import com.paperconnect.exception.KeywordException;

/**
 * Entry point class for the application
 */
public class PaperConnect implements EntryPoint {
	
	//List to store the papers currently in the table
	private ArrayList<PaperShort> papers = new ArrayList<PaperShort>();
	
	//Asynchronous PaperService object to make calls to the server 
	private PaperServiceAsync paperSvc = GWT.create(PaperService.class);
	private Label errorMsgLabel = new Label();

	/**
	 * Method for initializing and populating the HTML page
	 */
	public void onModuleLoad() {

		//Set the container used for displaying the sigma.js graph invisible
		RootPanel.get("graphContainer").setVisible(false);

		// Associate the errorMessageLabel with the HTML host page
		RootPanel.get("errorMessage").add(errorMsgLabel);

		//Initialize the JSNI method for the searchBar
		searchListener(this);
		
		//Initialize the JSNI method for the tableListener
		tableListener(this);
	}

	/**
	 * Method for adding the list of papers queried to the list of papers for the table 
	 * @param paperLs
	 */
	private void addPapers(ArrayList<PaperShort> paperLs) {
		papers.clear();
		papers = paperLs;
	}

	/**
	 * Method for retrieving a {@link Paper} from the server
	 * @param title - the title of the {@link PaperShort} selected
	 */
	private void retrievePaper(String title) {
		String id = "";
		//Look through the list of papers in the table
		for(PaperShort paper : papers) {
			//Check if this is the paper that was selected
			if(paper.getField(PaperFields.TITLE).equals(title)) {
				//Retrieve the id for this paper
				id = paper.getField(PaperFields.ID);
				System.out.println(id+" "+paper.getField(PaperFields.TITLE));
				break;
			}
		}
		
		// Initialize the service proxy
		if (paperSvc == null) {
			paperSvc = GWT.create(PaperService.class);
		}

		// Set up the callback object.
		AsyncCallback<Paper> callback = new AsyncCallback<Paper>() {
			public void onFailure(Throwable caught) {
				if(caught instanceof InvalidIdException) {
					//Set the errorMessage label text and make it visible
					errorMsgLabel.setText(((InvalidIdException) caught).getMessage());
					errorMsgLabel.setVisible(true);
				}
			}

			@Override
			public void onSuccess(Paper result) {
				//Initialize the HTML table for displaying the paper
				initPaperTable("paperTable", true);
				
				JSONObject obj = new JSONObject();
				
				//Initialize the JSON object with the fields that will be displayed
				obj.put("title", new JSONString(result.getField(PaperFields.TITLE)));
				obj.put("author", new JSONString(result.getField(PaperFields.AUTHOR)));
				obj.put("publishDate", new JSONString(result.getField(PaperFields.PUBLISH_DATE)));
				obj.put("abstract", new JSONString(result.getField(PaperFields.ABSTRACT)));
				
				//Add the paper to the HTML table
				addPaperInTable(obj, "paperTable", true);
				
				//Parse the JSONString for the graph into JSONObject that can be used by sigma.js
				JSONValue val = JSONParser.parseStrict(result.getField(PaperFields.TREE));
				JSONObject graphObj = val.isObject();
				
				//Set the container used for displaying the sigma.js graph visible
				RootPanel.get("graphContainer").setVisible(true);
				
				//Display the graph in the graphContainer using sigma.js
				displayGraph(graphObj, "graphContainer");
				
				//Set the errorMessage label to invisible
				errorMsgLabel.setVisible(false);
			}
		};

		// Make the call to retrieve the paper
		paperSvc.retrievePaper(id, callback);
	}

	/**
	 * Method for retrieving a list of {@link PaperShort} related to the inputed keyword
	 * @param keyword - the keyword inputed in the search bar
	 */
	private void retrievePaperLs(String keyword) {
		// Initialize the service proxy
		if (paperSvc == null) {
			paperSvc = GWT.create(PaperService.class);
		}

		// Set up the callback object.
		AsyncCallback<ArrayList<PaperShort>> callback = new AsyncCallback<ArrayList<PaperShort>>() {
			public void onFailure(Throwable caught) {
				if(caught instanceof KeywordException) {
					//Set the errorMessage label text and make it visible
					errorMsgLabel.setText(((KeywordException) caught).getMessage());
					errorMsgLabel.setVisible(true);
				}
			}

			public void onSuccess(ArrayList<PaperShort> result) {
				
				//Set the container used for displaying the sigma.js graph invisible
				RootPanel.get("graphContainer").setVisible(false);
				
				//Initialize the HTML table for displaying the list of papers
				initPaperTable("paperTable", false);
				
				//Add the papers from the query to the list of papers that will be in the table
				addPapers(result);
				
				//Add every paper in from the query into the table
				for (PaperShort paper : result) {
					JSONObject obj = new JSONObject();
					
					//Initialize the JSON object with the fields that will be displayed
					obj.put("title", new JSONString(paper.getField(PaperFields.TITLE)));
					obj.put("author", new JSONString(paper.getField(PaperFields.AUTHOR)));
					obj.put("publishDate", new JSONString(paper.getField(PaperFields.PUBLISH_DATE)));
					
					//Add the paper to the HTML table
					addPaperInTable(obj, "paperTable", false);
				}
				//Set click handlers for the rows of the HTML table
				addClickHandlersToTable("paperTable");
				
				//Set the errorMessage label to invisible
				errorMsgLabel.setVisible(false);
			}
		};

		// Make the call to retrieve the list of papers
		paperSvc.retrievePaperLs(keyword, callback);
	}

	/**
	 * JSNI method for calling the JavaScript method to initialize the HTML table, for either single or
	 * multiple paper configurations
	 * @param elementId - The id of the element in which to generate the table
	 * @param singlePaper - Whether the table configuration should for a single paper or multiple
	 */
	public native void initPaperTable(String elementId, boolean singlePaper) /*-{
		$wnd.initPaperTable(elementId, singlePaper);
	}-*/;

	/**
	 * JSNI method for calling the JavaScript method to draw the graph using sigma.js
	 * @param obj - JSONObject containing all the relevant data (nodes and edges) for the graph
	 * @param elementId - The id of the element in which to draw the graph
	 */
	public native void displayGraph(JSONObject obj, String elementId) /*-{
		$wnd.drawGraph(obj, elementId);
	}-*/;

	/**
	 * JSNI method for calling the JavaScript method to add a paper to the table
	 * @param obj - JSONObject containing the all the relevant data for the paper
	 * @param elementId - The id of the element in which the table exists
	 * @param singlePaper - Which configuration of the table is it adding the paper to 
	 */
	public native void addPaperInTable(JSONObject obj, String elementId, boolean singlePaper) /*-{
		$wnd.addPaperToTable(obj, elementId, singlePaper);
	}-*/;

	/**
	 * JSNI method for calling the JavaScript method to add clickHandlers for all the rows of the table
	 */
	public native void addClickHandlersToTable(String elementId) /*-{
		$wnd.addRowHandlers();
	}-*/;

	/**
	 * JSNI method calling the Java method retrievePaperLs from the client side in the event of a search bar query 
	 * @param pc - instance of this class
	 */
	public native void searchListener(PaperConnect pc) /*-{
		$wnd.onSearchEnter = function(keyword) {
			pc.@com.paperconnect.client.PaperConnect::retrievePaperLs(Ljava/lang/String;)(keyword);
		};
	}-*/;

	/**
	 * JSNI method calling the Java method retrievePaper from the client side in the event of a paper being selected
	 * @param pc - instance of this class
	 */
	public native void tableListener(PaperConnect pc) /*-{
		$wnd.onTableClick = function(title) {
			pc.@com.paperconnect.client.PaperConnect::retrievePaper(Ljava/lang/String;)(title);
		};
	}-*/;
}

File: PaperConnect/war/paperconnect.js
//Function for initialize the HTML table, either for displaying single paper or list of papers
function initPaperTable(elementId, flag) {

	// Clear the table
	document.getElementById(elementId).innerHTML = "";

	// Create a table as a list 
	var tableHeader = document.createElement('LI');
	
	//Create the first row of the table, which is the table header
	var title = document.createElement("DIV");
	var author = document.createElement("DIV");
	var publishDate = document.createElement("DIV");
	var abstract = document.createElement("DIV");

	//Add className attribute field for referencing from the css
	tableHeader.className = "table-header";
	title.className = "col col-1";
	author.className = "col col-2";
	publishDate.className = "col col-3";
	abstract.className = "col col-4";

	//Set the div elements to the corresponding table headers
	title.innerHTML = "TITLE";
	author.innerHTML = "AUTHOR";
	publishDate.innerHTML = "PUBLISH DATE";
	abstract.innerHTML = "ABSTRACT";
	
	//Append the table headers to the table
	tableHeader.appendChild(title);
	tableHeader.appendChild(author);
	tableHeader.appendChild(publishDate);
	
	//If the configuration is for a single paper add an additional column for the abstract
	if(flag){
		tableHeader.appendChild(abstract);
	}
	
	//Append the table to the element in which it needs to embedded
	document.getElementById(elementId).appendChild(tableHeader);
}

//Function for adding papers to the table, either for a single paper configuration or list of papers
function addPaperToTable(obj, elementId, flag) {
	var paper = JSON.parse(obj);
	
	//Create a table row for embedding the data
	var tableRow = document.createElement('LI');
	
	//Div elements representing the table row data
	var paperTitle = document.createElement("DIV");
	var paperAuthor = document.createElement("DIV");
	var paperPublishDate = document.createElement("DIV");
	var paperAbstract = document.createElement("DIV");
	
	//Create an attribute for the data for referencing from the css
	var data_label_title = document.createAttribute("data-label");
	var data_label_author = document.createAttribute("data-label");
	var data_label_publishDate = document.createAttribute("data-label");
	var data_label_abstract = document.createAttribute("data-label");
	
	//Add the className attribute field for referencing from the css
	tableRow.className = "table-row";
	paperTitle.className = "col col-1";
	paperAuthor.className = "col col-2";
	paperPublishDate.className = "col col-3";
	paperAbstract.className = "col col-4";

	//Set the value of the attribute and add it to the appropriate div
	data_label_title.value = "Title";
	paperTitle.setAttributeNode(data_label_title);

	//Set the value of the attribute and add it to the appropriate div
	data_label_author.value = "Author";
	paperAuthor.setAttributeNode(data_label_author);

	//Set the value of the attribute and add it to the appropriate div
	data_label_publishDate.value = "Publish Date";
	paperPublishDate.setAttributeNode(data_label_publishDate);

	//Set the value of the attribute and add it to the appropriate div
	data_label_abstract.value = "Abstract";
	paperAbstract.setAttributeNode(data_label_abstract);
	
	//Set the value of the text to be displayed in the table and append it to the table row
	paperTitle.innerHTML = paper.title;
	tableRow.appendChild(paperTitle);
	
	//Set the value of the text to be displayed in the table and append it to the table row
	paperAuthor.innerHTML = paper.author;
	tableRow.appendChild(paperAuthor);

	//Set the value of the text to be displayed in the table and append it to the table row
	paperPublishDate.innerHTML = paper.publishDate;
	tableRow.appendChild(paperPublishDate);
	
	//If the configuration is for list of papers add the additional field
	if(flag){
		//Set the value of the text to be displayed in the table and append it to the table row
		paperAbstract.innerHTML = paper.abstract;
		tableRow.appendChild(paperAbstract);	
	}
	document.getElementById(elementId).appendChild(tableRow);
}

//Function for addClickHandler to every row in the table
function addRowHandlers(elementId) {
	//Retrieve the table by elementId
	var table = document.getElementById(elementId);
	
	//Retrieve the rows by tag
	var rows = table.getElementsByTagName("LI");
	
	//For every row in the table set a clickHandler to it
	for (i = 1; i < rows.length; i++) {
		//Get the current row by index
		var currentRow = rows[i];
		
		//Retrieve the data in this row by tag
		var divTags = rows[i].getElementsByTagName("DIV");
		
		//Get the first data element in this row (TITLE)
	    var titleData = divTags[0].innerHTML;
	    
	    //Create a clickHandler
	    var clickHandler = function(title) {
	        return function() {
	        	//Call the Java method onTableClick and pass in the title of the paper selected
	        	window.onTableClick(title);
	        };
	    };
	    
	    //Set the onClick method for the row to the clickHandler  
	    currentRow.onclick = clickHandler(titleData);
	}
}

function drawGraph(obj, elementId) {
	
	//Clear the graph
	document.getElementById(elementId).innerHTML = "";
	
	 // Initialize sigma:
	var s = new sigma({
		renderer : {
			container : document.getElementById(elementId),
			type : 'canvas'
		},
		settings : {}
	});
	
	s.graph.clear();
	
	//Create a JSON object store the nodes and edges to be used by sigma
	var graph = {
		nodes : [],
		edges : []
	};
	
	var jsonGraph = JSON.parse(obj);
	
	//For every node in the jsonGraph read the data and put it in graph
	for (i = 0; i < jsonGraph.nodes.length; i++) {
		//Push a new node containing the necessary fields into the nodes list
		graph.nodes.push({
			id : jsonGraph.nodes[i].id,
			label : jsonGraph.nodes[i].label,
			x : jsonGraph.nodes[i].x,
			y : jsonGraph.nodes[i].y,
			size : jsonGraph.nodes[i].size,
			color: '#008CC2'
		});
	}
	
	//For every edge in the jsonGraph read the data and put in graph
	for (i = 0; i < jsonGraph.edges.length; i++) {
		//Push a new edge containing the necessary fields into the edges list
		graph.edges.push({
			id : jsonGraph.edges[i].id,
			source : jsonGraph.edges[i].source,
			target : jsonGraph.edges[i].target,
			 color: '#DC143C',
		});
	}

	// Load the graph in sigma
	s.graph.read(graph);
	
	// Ask sigma to draw it
	s.refresh();
}